runtime/src/main/java/org/corfudb/runtime/DistributedCompactor.java

```diff
@@ -0,0 +1,558 @@
+                endTxn.commit();
+                break;
+            } catch (RuntimeException re) {
+                if (isCriticalRuntimeException(re, retry, maxRetries)) {
+                    return checkpointFailedError; // stop on non-retryable exceptions
+                }
+            } catch (Throwable t) {
+                log.error("Unexpected exception encountered while unlocking my checkpoint table, ", t);
+                log.error("StackTrace : {}", t.getStackTrace());
+                isSuccess = checkpointFailedError;
+                break;
+            }
+        }
+        return isSuccess;
+    }
+
+    /**
+     * @param re         - the exception this method is called on
+     * @param retry      - the number of times retries have been done
+     * @param maxRetries - max number of times retries need to happen
+     * @return - True if we should stop & return. False if we can retry!
+     */
+    private boolean isCriticalRuntimeException(RuntimeException re, int retry, int maxRetries) {
+        log.trace("checkpointer: encountered an exception on attempt {}/{}.",
+                retry, maxRetries, re);
+
+        if (retry == maxRetries - 1) {
+            log.error("checkpointer: retry exhausted.", re);
+            return true;
+        }
+
+        if (re instanceof NetworkException || re.getCause() instanceof TimeoutException) {
+            try {
+                TimeUnit.MILLISECONDS.sleep(DistributedCompactor.CONN_RETRY_DELAY_MILLISEC);
+            } catch (InterruptedException e) {
+                log.error("Interrupted in network retry sleep");
+                return true;
+            }
+        }
+        if (re instanceof WrongClusterException) {
+            log.error("Wrong cluster exception hit! stopping right away!");
+            return true;
+        }
+        return false; // it is ok to retry a few times on network timeouts
+    }
+
+    public static TableName getTableName(Table<Message, Message, Message> table) {
+        String fullName = table.getFullyQualifiedTableName();
+        return TableName.newBuilder()
+                .setNamespace(table.getNamespace())
+                .setTableName(fullName.substring(fullName.indexOf("$") + 1))
+                .build();
+    }
+
+    /**
+     * In the global checkpoint map we examine if there is a special "freeze token"
+     * The sequence part of this token is overloaded with the timestamp
+     * when the freeze was requested.
+     * Now checkpointer being a busybody has limited patience (2 hours)
+     * If the freeze request is within 2 hours it will honor it and step aside.
+     * Otherwise it will angrily remove the freezeToken and continue about
+     * its business.
+     *
+     * @return - true if checkpointing should be skipped, false if not.
+     */
+    public static boolean isCheckpointFrozen(CorfuStore corfuStore, final Table<StringKey, TokenMsg, Message> chkptMap) {
```

Comment:
It seems like the main purpose of this method is to enable some ad-hoc process to pause the compaction service in a distributed manner. In that case, only a signal is needed and overloading a logical sequence number with actual data timestamp is unnecessarily complicated, and also incorrect (in the case of a clock skew).

Reply:
This method is the contract we have with LogReplicator module which we cannot break. so this logic had to be retained.